src/tpm/android/utils.rs

```diff
@@ -194,3 +196,32 @@ pub fn get_signature_algorithm(mode: EncryptionMode) -> Result<String, SecurityM
         }
     }
 }
+
+pub fn get_iv_len(cipher: BlockCiphers) -> Result<usize, SecurityModuleError> {
+    Ok(match cipher {
+        BlockCiphers::Aes(mode, _) => match mode {
+            SymmetricMode::Gcm => 12,
+            SymmetricMode::Ccm => 16,
+            SymmetricMode::Ecb => 16,
+            SymmetricMode::Cbc => 16,
+            SymmetricMode::Cfb => 16,
+            SymmetricMode::Ofb => 16,
+            SymmetricMode::Ctr => 16,
+        },
+        BlockCiphers::TripleDes(_) => 8,
+        BlockCiphers::Des | BlockCiphers::Rc2(_) | BlockCiphers::Camellia(_, _) => {
+            return Err(TpmError::UnsupportedOperation("not supported".to_owned()).into())
+        }
+    })
+}
+
+pub fn store_iv(mut data: Vec<u8>, mut iv: Vec<u8>) -> Vec<u8> {
+    iv.append(&mut data);
```

Comment:
Refer to https://doc.rust-lang.org/std/vec/struct.Vec.html#method.append.

in cases where IV is stored more than once you will run into problems. 
Remember to clear iv before doing the append. See https://doc.rust-lang.org/std/vec/struct.Vec.html#method.clear.

Reply:
As the data vector is moved into the function, it statically cannot be used after calling store_iv.
Clearing the iv vector would remove the purpose of the function, as the iv vector already contains the IV when the function is called.

Reply:
Let's look at your code line-by-line. Firstly, this is a public method. Effectively this means that others may also make use of this method to store iv in a vector. The method name indicates that IV is stored in a destination vector. You then append everything in the data vector to the iv vector. You then return iv. This is fine, if iv is empty, but will not work if iv has been used previously.

pub fn store_iv(mut data: Vec<u8>, mut iv: Vec<u8>) -> Vec<u8> {
    iv.append(&mut data);
    return iv;
}

In mod.rs you did the following:
let iv = cipher.getIV(&env).err_internal()?;
                let encrypted = cipher.doFinal(&env, data.to_vec()).err_internal()?;
                store_iv(encrypted, iv)

Granted, for this specific instance, your implementation of store_iv will work. However, as pointed out, this is a public method that others can potentially also use. Should anyone else use this in a different fashion, then they may run into IV issues if the destination vector is re-used.

Reply:
Could you please provide an example of how this function could be misused?

I don't quite understand why the IV vector needs to be empty. In the code snippet you brought up, the IV vector already contains the IV before the call to `store_iv`.
```rust
let result_vector = {
    let iv = cipher.getIV(&env).err_internal()?;
    // the vector iv now contains the generated iv

    let encrypted = cipher.doFinal(&env, data.to_vec()).err_internal()?;  
    // the vector encrypted now contains the encrypted data

    store_iv(encrypted, iv)    
    // the new vector returned by `store_iv` is returned as the result of the block and assigned to result_vector
    // iv and encrypted were consumed by store_iv and are no longer accessible
}
```

To your concerns about the function being public:
While the function itself is declared as public, the containing module is declared as `pub(crate)` and is not accessible outside of the crate and not part of the public API surface. Is your concern that the function could be misused by other internal parts of the library ?